src/Castle.Facilities.AspNetCore/CastleWindsorRegistrationExtensions.cs

```diff
@@ -29,44 +29,185 @@ namespace Castle.Facilities.AspNetCore
+		/// <param name="lifestyleType">The lifestyle of the tag helpers. Defaults to LifestyleType.Scoped.</param>
+		/// <returns>CastleWindsorRegistrationOptions as a fluent interface</returns>
+		public CastleWindsorRegistrationOptions RegisterTagHelpers(Assembly tagHelpersAssembly, LifestyleType lifestyleType = LifestyleType.Scoped)
+		{
+			tagHelpersAlreadyRegistered = true;
+			container.Register(Classes.FromAssemblyInThisApplication(tagHelpersAssembly).BasedOn<TagHelper>().Configure(x => x.LifeStyle.Is(lifestyleType)));
+			return this;
+		}
+
+		/// <summary>
+		/// Use this method to customise the registration/lifestyle of view components.
+		/// </summary>
+		/// <param name="viewComponentsAssembly">Assembly where the view components are defined. Defaults to Assembly.GetCallingAssembly().</param>
+		/// <param name="lifestyleType">The lifestyle of the view components. Defaults to LifestyleType.Scoped.</param>
+		/// <returns>CastleWindsorRegistrationOptions as a fluent interface</returns>
+		public CastleWindsorRegistrationOptions RegisterViewComponents(Assembly viewComponentsAssembly, LifestyleType lifestyleType = LifestyleType.Scoped)
+		{
+			viewComponentsAlreadyRegistered = true;
+			container.Register(Classes.FromAssemblyInThisApplication(viewComponentsAssembly).BasedOn<ViewComponent>().Configure(x => x.LifeStyle.Is(lifestyleType)));
+			return this;
+		}
+
+	}
+
 	public static class CastleWindsorRegistrationExtensions
 	{
 		/// <summary>
 		/// Sets up framework level activators for Controllers, TagHelpers and ViewComponents and adds additional sub dependency resolvers
 		/// </summary>
 		/// <param name="services">ASP.NET Core service collection from Microsoft.Extensions.DependencyInjection</param>
 		/// <param name="container">Windsor container which activators call resolve against</param>
-		public static void AddCastleWindsor(this IServiceCollection services, IWindsorContainer container)
+		/// <param name="configure">Configuration options for controlling registration and lifestyles of controllers, tagHelpers and viewComponents</param>
+		public static void AddCastleWindsor(this IServiceCollection services, IWindsorContainer container, Action<CastleWindsorRegistrationOptions> configure = null)
 		{
-			services.AddRequestScopingMiddleware(container.BeginScope);
-			services.AddCustomControllerActivation(container.Resolve);
-			services.AddCustomTagHelperActivation(container.Resolve);
-			services.AddCustomViewComponentActivation(container.Resolve);
-			container.Kernel.Resolver.AddSubResolver(new LoggerDependencyResolver(services));
-			container.Kernel.Resolver.AddSubResolver(new FrameworkConfigurationDependencyResolver(services));
+			AddFrameworkResolversToWindsor(services, container);
+			InstallFrameworkIntegration(services, container);
+
+			var options = new CastleWindsorRegistrationOptions(container, services);
+			(configure ?? ((o) => { }))(options);
```

Comment:
This reads really poorly and would invoke that lambda even though it does nothing, could you change this to an if statement with a null check.

Reply:
I will do. It is just lambda parameter based null coalescence to avoid null reference exceptions. Is this bad? Can I still use null coalescence for invalid arguments?

Reply:
> Can I use null coalescence for invalid arguments?

Yes, definitely, those look great to handle the throw exception case.

> It is just lambda parameter based null coalescence to avoid null reference exceptions. Is this bad?

Other than readability my concern is that doing it this way actually runs more code at runtime and that empty lambda gets emitted as a compiler named method.

Reply:
Done

Reply:
`configure?.Invoke(options)` is much better, and it happens that it is the exact same IL with the other syntax.